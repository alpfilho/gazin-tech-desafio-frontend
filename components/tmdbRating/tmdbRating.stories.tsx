import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { TmdbRating } from './tmdbRating';

export default {
	title: 'Tmdb Rating',
	component: TmdbRating,
	argTypes: {
		rating: {
			control: { type: 'select' },
			options: [10, 6.7, 4.5, 0, 8.5],
			defaultValue: 0
		}
	}
} as ComponentMeta<typeof TmdbRating>;

export const Default: ComponentStory<typeof TmdbRating> = ({ rating }) => (
	<TmdbRating rating={rating || 0} />
);
