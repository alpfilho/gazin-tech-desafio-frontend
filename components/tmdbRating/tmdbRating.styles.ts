import styled from 'styled-components';

export const RatingContainer = styled.div`
	display: flex;
	flex-direction: row;
	align-items: center;
	justify-content: center;
	user-select: none;
`;

export const SvgContainer = styled.div`
	display: flex;
	justify-content: center;
	align-items: center;

	svg {
		height: 100%;
		width: auto;
	}
`;

export const StarContainer = styled(SvgContainer)`
	height: 2em;
`;

export const RatingText = styled.div`
	display: flex;
	justify-content: center;
	align-items: center;

	white-space: nowrap;

	font-family: 'Poppins', sans-serif;
	font-weight: 700;
	font-size: 2.75em;

	@media screen and (max-width: 768px) {
		font-size: 2em;
	}
`;

export const Rating = styled.div`
	width: 2em;
	display: flex;
	justify-content: center;
	text-align: center;
	align-items: center;
`;

export const RatingLimit = styled.div`
	width: 2.125em;
	font-size: 80%;
	display: flex;
	justify-content: flex-start;
	text-align: left;
	align-items: center;
	color: #ffb800;
`;

export const TmbdLogoContainer = styled.div`
	width: 5em;
`;
