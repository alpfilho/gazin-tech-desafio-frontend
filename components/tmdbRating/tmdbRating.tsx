import React, { useEffect, useRef } from 'react';
import { useMotionValue, animate } from 'framer-motion';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faStar } from '@fortawesome/free-solid-svg-icons';

import {
	TmbdLogoContainer,
	RatingContainer,
	StarContainer,
	RatingText,
	Rating,
	RatingLimit
} from './tmdbRating.styles';
import { TmdbLogo } from 'components/tmdbLogo';

interface TmdbRatingI {
	rating?: number;
}

export const TmdbRating: React.FC<TmdbRatingI> = ({ rating = 0 }) => {
	const ratingValueContentRef = useRef<HTMLDivElement>(null);
	const ratingValue = useMotionValue(0);

	useEffect(() => {
		if (ratingValueContentRef.current) {
			ratingValueContentRef.current.innerHTML = '0';
		}

		ratingValue.onChange((value) => {
			if (ratingValueContentRef.current) {
				const ratingValue = value.toFixed(1);
				const ratingValueFloat = parseFloat(ratingValue);

				ratingValueContentRef.current.innerHTML = `${
					ratingValueFloat >= 10 ? 10 : ratingValueFloat <= 0 ? 0 : ratingValue
				}`;
			}
		});

		return () => {
			ratingValue.clearListeners();
		};
	}, [ratingValue]);

	useEffect(() => {
		const animationControl = animate(ratingValue, rating, {
			type: 'spring',
			stiffness: 300,
			damping: 50,
			restDelta: 0.1
		});
		return () => {
			animationControl.stop();
		};
	}, [rating, ratingValue]);

	return (
		<RatingContainer>
			<StarContainer>
				<FontAwesomeIcon icon={faStar} color="#FFB800" />
			</StarContainer>

			<RatingText>
				<Rating ref={ratingValueContentRef} />
				<RatingLimit>/ 10</RatingLimit>
			</RatingText>

			<TmbdLogoContainer>
				<TmdbLogo type="primary-short" />
			</TmbdLogoContainer>
		</RatingContainer>
	);
};
