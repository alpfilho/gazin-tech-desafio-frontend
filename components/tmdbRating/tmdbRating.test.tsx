import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';

import * as stories from './tmdbRating.stories';

const { Default } = composeStories(stories);

describe('Tmdb Rating', () => {
	it('render a Tmdb rating', () => {
		render(<Default />);

		const ratingText = screen.getByText('0');
		expect(ratingText).toBeInTheDocument();
	});

	it('Tmdb rating render unchanged', () => {
		const { container } = render(<Default />);

		expect(container).toMatchSnapshot();
	});
});
