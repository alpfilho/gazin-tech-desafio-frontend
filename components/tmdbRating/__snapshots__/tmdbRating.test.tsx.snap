// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tmdb Rating Tmdb rating render unchanged 1`] = `
<div>
  <div
    class="RatingContainer-sc-1g1w7oy-0 eRwBjt"
  >
    <div
      class="SvgContainer-sc-1g1w7oy-1 StarContainer-sc-1g1w7oy-2 evdtLQ gZuYIb"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-star "
        color="#FFB800"
        data-icon="star"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M381.2 150.3L524.9 171.5C536.8 173.2 546.8 181.6 550.6 193.1C554.4 204.7 551.3 217.3 542.7 225.9L438.5 328.1L463.1 474.7C465.1 486.7 460.2 498.9 450.2 506C440.3 513.1 427.2 514 416.5 508.3L288.1 439.8L159.8 508.3C149 514 135.9 513.1 126 506C116.1 498.9 111.1 486.7 113.2 474.7L137.8 328.1L33.58 225.9C24.97 217.3 21.91 204.7 25.69 193.1C29.46 181.6 39.43 173.2 51.42 171.5L195 150.3L259.4 17.97C264.7 6.954 275.9-.0391 288.1-.0391C300.4-.0391 311.6 6.954 316.9 17.97L381.2 150.3z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="RatingText-sc-1g1w7oy-3 dKfZRU"
    >
      <div
        class="Rating-sc-1g1w7oy-4 iJEgzH"
      >
        0
      </div>
      <div
        class="RatingLimit-sc-1g1w7oy-5 kJwsdo"
      >
        / 10
      </div>
    </div>
    <div
      class="TmbdLogoContainer-sc-1g1w7oy-6 cczYB"
    >
      <svg
        data-testid="tmdb-logo-primary-short"
        viewBox="0 0 190.24 81.52"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <lineargradient
            gradientUnits="userSpaceOnUse"
            id="imdb-logo-gradient-1"
            x2="190.24"
            y1="40.76"
            y2="40.76"
          >
            <stop
              offset="0"
              stop-color="#90cea1"
            />
            <stop
              offset=".56"
              stop-color="#3cbec9"
            />
            <stop
              offset="1"
              stop-color="#00b3e5"
            />
          </lineargradient>
        </defs>
        <g
          data-name="Layer 2"
        >
          <path
            d="M105.67 36.06h66.9a17.67 17.67 0 0 0 17.67-17.66A17.67 17.67 0 0 0 172.57.73h-66.9A17.67 17.67 0 0 0 88 18.4a17.67 17.67 0 0 0 17.67 17.66Zm-88 45h76.9a17.67 17.67 0 0 0 17.67-17.66 17.67 17.67 0 0 0-17.67-17.67h-76.9A17.67 17.67 0 0 0 0 63.4a17.67 17.67 0 0 0 17.67 17.66Zm-7.26-45.64h7.8V6.92h10.1V0h-28v6.9h10.1Zm28.1 0h7.8V8.25h.1l9 27.15h6l9.3-27.15h.1V35.4h7.8V0H66.76l-8.2 23.1h-.1L50.31 0h-11.8Zm113.92 20.25a15.07 15.07 0 0 0-4.52-5.52 18.57 18.57 0 0 0-6.68-3.08 33.54 33.54 0 0 0-8.07-1h-11.7v35.4h12.75a24.58 24.58 0 0 0 7.55-1.15 19.34 19.34 0 0 0 6.35-3.32 16.27 16.27 0 0 0 4.37-5.5 16.91 16.91 0 0 0 1.63-7.58 18.5 18.5 0 0 0-1.68-8.25ZM145 68.6a8.8 8.8 0 0 1-2.64 3.4 10.7 10.7 0 0 1-4 1.82 21.57 21.57 0 0 1-5 .55h-4.05v-21h4.6a17 17 0 0 1 4.67.63 11.66 11.66 0 0 1 3.88 1.87A9.14 9.14 0 0 1 145 59a9.87 9.87 0 0 1 1 4.52 11.89 11.89 0 0 1-1 5.08Zm44.63-.13a8 8 0 0 0-1.58-2.62 8.38 8.38 0 0 0-2.42-1.85 10.31 10.31 0 0 0-3.17-1v-.1a9.22 9.22 0 0 0 4.42-2.82 7.43 7.43 0 0 0 1.68-5 8.42 8.42 0 0 0-1.15-4.65 8.09 8.09 0 0 0-3-2.72 12.56 12.56 0 0 0-4.18-1.3 32.84 32.84 0 0 0-4.62-.33h-13.2v35.4h14.5a22.41 22.41 0 0 0 4.72-.5 13.53 13.53 0 0 0 4.28-1.65 9.42 9.42 0 0 0 3.1-3 8.52 8.52 0 0 0 1.2-4.68 9.39 9.39 0 0 0-.55-3.18Zm-19.42-15.75h5.3a10 10 0 0 1 1.85.18 6.18 6.18 0 0 1 1.7.57 3.39 3.39 0 0 1 1.22 1.13 3.22 3.22 0 0 1 .48 1.82 3.63 3.63 0 0 1-.43 1.8 3.4 3.4 0 0 1-1.12 1.2 4.92 4.92 0 0 1-1.58.65 7.51 7.51 0 0 1-1.77.2h-5.65Zm11.72 20a3.9 3.9 0 0 1-1.22 1.3 4.64 4.64 0 0 1-1.68.7 8.18 8.18 0 0 1-1.82.2h-7v-8h5.9a15.35 15.35 0 0 1 2 .15 8.47 8.47 0 0 1 2.05.55 4 4 0 0 1 1.57 1.18 3.11 3.11 0 0 1 .63 2 3.71 3.71 0 0 1-.43 1.92Z"
            data-name="Layer 1"
            fill="url(#imdb-logo-gradient-1)"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
